List Versions

irq0 opened the pull request:
Reimplementation of list versions and integration into bucket->list().
Leverage SQL (instead of get_all) and support pagination.
Populate object owner information. 

Add unit tests for corner cases, as direct s3-tests coverage is weaker than the regular list coverage. Indirectly lifecycle tests rely heavily on list versions. 

Fixes: https://github.com/aquarist-labs/s3gw/issues/647
Fixes: https://github.com/aquarist-labs/s3gw/issues/644
Epic: https://github.com/aquarist-labs/s3gw/issues/256

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

Commit:
rgw/sfs: Query-based list versions

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/sfs: Integrate new version listing into SFS bucket

Replace and remove previous get_all-based code with
SQLiteList::versions. Use the same delim logic as we use for listing
regular objects

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/sfs: Add bucket get_owner

Add SQLiteBuckets::get_owner to get the owner id and display name for
use with listings.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/sfs: Set ownership info in bucket listings

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
ci: Enable more s3 tests

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

jecluis commented on src/rgw/driver/sfs/bucket.cc:
```diff
@@ -148,149 +147,96 @@ int SFSBucket::list(
+                                 get_bucket_id(), params.prefix, start_with,
+                                 max, results.objs, &results.is_truncated
+                             )) ||
+      list.objects(
```
Maybe I'm missing something here, but won't this be running the `list.objects()` even if we `want_list_versions` but fail at `list.versions()` ? Is that meant by design?

irq0 replied:
Good catch. Tired to be clever..

jecluis commented on src/rgw/driver/sfs/bucket.cc:
```diff
@@ -148,149 +147,96 @@ int SFSBucket::list(
+    );
+
+    // Is there actually more after the rolled up common prefix? We
+    // can't tell form the original object query. Ask again for
```
s/form/from/

jecluis commented on src/rgw/driver/sfs/bucket.cc:
```diff
@@ -148,149 +147,96 @@ int SFSBucket::list(
-      }
+  if (!params.delim.empty()) {
+    std::vector<rgw_bucket_dir_entry> new_results;
+    list.roll_up_common_prefixes(
+        params.prefix, params.delim, results.objs, results.common_prefixes,
+        new_results
+    );
```
From what I can tell, won't this essentially filter the `result.objs` contents? If that's the case, isn't it possible that we will end up with less than `max` entries in the resulting `results.objs = new_results` (later in the function), even if the results have been truncated during the listing performed above?

irq0 replied:
I'd say it "compresses" `result.objs`. Multiple objs may become one common prefix.

We return less than `max` objs if len(common prefixes) > 0. 

I could not find a definitive answer if `max` applies to just objects returned or includes common prefixes. 
However, the docs said it is OK to return less than max. Important for proper pagination is the `next_marker`. Clients must use that to iterate a bucket - there is no start after 1000 objects.

To make the queries simpler this interprets `max` as `max(objects with search criteria)` and common prefixes as just a way to compress results. If we were to maximize results we would need more complex common prefix searches to find up to max of them. I'd say this is future work if we find that it would improve performance of clients doing lots of delimiter lookups. The less we return the more requests a client has to do. 

Fortunately, there are s3 tests that exercise the pagination and common prefix bits. They helped a lot and I hope that I did not just fit their interpretation ;)

jecluis commented on src/rgw/driver/sfs/sqlite/sqlite_buckets.cc:
```diff
@@ -42,6 +45,22 @@ std::optional<DBOPBucketInfo> SQLiteBuckets::get_bucket(
+    return std::nullopt;
+  }
+  const auto row = rows[0];
+  return std::make_pair(std::get<0>(row), std::get<1>(row).value());
```
feels like the either the first `get<>()` is missing a `.value()`, or the `.value()` on the second `get<>()` should not be there.

irq0 replied:
Thanks for pointing this out. It is actually neither. The second is an optional and value() should have been a value_or("")

jecluis commented on src/rgw/driver/sfs/sqlite/sqlite_list.cc:
```diff
@@ -78,7 +78,104 @@ bool SQLiteList::objects(
+          order_by(&DBVersionedObject::id).desc()
+      ),
+      limit(query_limit)
+  );
```
okay, so this query made my ears bleed a bit, but it looks sound. I'm trusting it blindly.

jecluis reviewed (changes requested):
more of a "request for clarification" than a "request changes". It looks good though.

Commit:
rgw/sfs: Integrate new version listing into SFS bucket

Replace and remove previous get_all-based code with
SQLiteList::versions. Use the same delim logic as we use for listing
regular objects

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>
